### DISCUSSION TITLE
Efficient search / edit / continue search workflow?

### DISCUSSION THREAD

**Discussion by shmerl** [ibhagwan/fzf-lua#1126] [Q&A] [answered]
Is there a way to

1. Search for something (let's say rg / grep search in files) which opens an fzf lua UI with found entries.
2. Iterate over found entries - open it in some separate place (a tab may be?)
3. Modify that found place
4. Continue iteration over the above until you finish all such modifications

So far if I try to access an entry to edit it (by pressing Enter) it jumps there but closes the search UI. So I need to re-run search after each edit, or alternatively I need to copy the file / line from the found result and open it explicitly somewhere else (like separate editor instance?).

Is there a way to do it better?

**Comment by ibhagwan** [comment, marked as answer]
There is a much better workflow, a few options:

- First open `live_grep` and enter a wide capturing regex to narrow down the results 
- Then press `ctrl-g` to toggle to fuzzy search so you’re already working on a smaller set
- Closed the interface? No problem, run `:FzfLua grep resume=true` and continue where you left off, both search regex and fuzzy query will be resumed so you can continue the `ctrl-g` back and forth

To further improve on this add a history file to grep and you’ll be able to go back/next on your searches:
https://github.com/ibhagwan/fzf-lua/wiki#how-do-i-setup-input-history-keybinds

Another refinement options is to select all entries you’re interested in and press enter, by default multiple entries will go to quick fix, you can then go back and forth the quickfix wirh `:cnext` and `:cprev` (I map these to `]q` and `[q`) and also explore these with `:FzfLua quickfix`.

**Reply by shmerl** [reply]
Thanks, I can try some of that. How can I pass resume = true if I'm using more explicit command something like:

```
:lua require'fzf-lua'.live_grep_native({ cmd = "rg <some rg customization params>" })
```

**Reply by ibhagwan** [reply]
```lua
:lua require'fzf-lua'.live_grep_native({ cmd = "rg <some rg customization params>" , resume = true })
```

**Reply by shmerl** [reply]
Thanks! That works mostly, except position of the search entry in the interface after you do resume resets to the first one. But still it's a big improvement over starting with an empty search.

**Comment by ibhagwan** [comment]
> Thanks! That works mostly, except position of the search entry in the interface after you do resume resets to the first one.

That unfortuntaly is not possible due to fzf limitations being an external process.

**Comment by shmerl** [comment]
> Another refinement options is to select all entries you’re interested in and press enter,

How do  you select several entries for it?

**Reply by ibhagwan** [reply]
By default with `tab`.

**Reply by shmerl** [reply]
Neat! Quickfix list is useful (never used it before).

**Reply by ibhagwan** [reply]
It’s very useful, for example for search and replace, you can use `:cdo` or `:cfdo`, this is how I use it, see my vim-cheatsheet:
> Search for `cdo`
https://github.com/ibhagwan/vim-cheatsheet?tab=readme-ov-file#search-replace

**Reply by shmerl** [reply]
It could be nice to be able to delete quickfix entries from the list from the fzf-lua UI itself. Sort of like dap breakpoints can be deleted?

**Reply by ibhagwan** [reply]
It is possible, select any entries you’re after and enter will create a new list, you can also presss ctrl-a for toggle-all for the inverse list.

**Reply by shmerl** [reply]
Where does ctrl+a work? It doesn't seem to do anything when in `:FzfLua quickfix`

I think being able to delete a quickfix list entry from there (including all entries) would still be neat.

**Reply by ibhagwan** [reply]
Might be alt-a, press F1 to see the help.

I don’t intend to implement quick fix deletion.

**Reply by shmerl** [reply]
Yeah, it was Alt+a. Btw, Just in case anyone needs, I found a plugin that handles deletion of quickfix list entry with using `dd` in it:

https://github.com/TamaMcGlinn/quickfixdd

**Reply by ibhagwan** [reply]
You can also quite easily create an action mapped to `toggle-all+accept`, it will create a new list from the non selected items which is essentially a delete.

**Reply by shmerl** [reply]
I might be missing understanding this idea about toggle all. If let's say I select 3 out of 10 entries in the grep search, toggle will make it into different 7 out of 10. And toggling anything in quickfix list itself didn't really affect things.

**Reply by ibhagwan** [reply]
Toggle+accept will create a new list with the 7 items (post toggle) - isn’t that essentially a deletion of the 3 selected?

**Reply by ibhagwan** [reply]
Add this to your quick actions and test:

```lua
require'fzf-lua'.setup {
  quickfix = {
    actions = { ['ctrl-x'] = { fn = require'fzf-lua'.actions.file_sel_to_qf , prefix = 'toggle-all+' } },
  }
}
```

**Reply by shmerl** [reply]
It doesn't really work that way for me when I reach 2 entries. I.e. for example selecting 2 out of 3 entries in the fzf's quickfix list deletes 1, but selecting 1 out of 2 doesn't delete it. And then, how do you even delete the last one (1 out of 1)?

**Reply by ibhagwan** [reply]
The default accept action sends any selection over 2 to a new quick fix list, therefore when there is one entry it will just open the file and not create a new list, I posted code for you above that always sends the invert selection to a new quick fix list (therefore deleting the selection).

**Reply by ibhagwan** [reply]
Btw, if you want the list interface to remain open add `reload = true` to the action parameters.

**Reply by ibhagwan** [reply]
The action name is `file_sel_to_qf`:
https://github.com/ibhagwan/fzf-lua/blob/743647f639a83e41e283d2d7daa03a85e1fbf951/lua/fzf-lua/defaults.lua#L103

**Reply by ibhagwan** [reply]
If you want finer control, you can write your own custom action, you can start with the code of the original action:
https://github.com/ibhagwan/fzf-lua/blob/743647f639a83e41e283d2d7daa03a85e1fbf951/lua/fzf-lua/actions.lua#L193-L234

**Reply by shmerl** [reply]
Ctrl+x didn't seem to work or do anything when I added that setup. Even without the action, I just don't understand what is supposed to happen. Here is what I was doing manually, I might miss some steps you have in mind:

1. Find something with fzf grep.
2. Use tab on a few entries and press enter - it creates a quickfix list.
3. Open quicfkix list itself in with FzfLua quickfix.
4. Select a subset of it and press enter.

4 already fails if you have only two entries and want to select one. And it also means you first need to select all and then unselect some. So that didn't work if I unselected all for example. I.e. doesn't do anything as "delete all" idea.

So it's seems way more straightforward to just do dd in the actual list like above plugin.

**Reply by ibhagwan** [reply]
The code I wrote (which I didn’t test as I’m on mobile ATM), adds a ctrl-x action to `:FzfLua quicfix` that inverts selection and creates a new list from that (essentially deletion).

I may have missed or mistyped some stuff as I’m on mobile but if you search the issues / read the wiki about similar toggles you’ll understand how to make it work.

If you feel better with a dedicated plugin for that by all means stay with it but with a little bit of effort and code you can make any customization happen with this plugin.

**Reply by shmerl** [reply]
I can take a closer look or see if I can implement deletion directly, to avoid using the inversion idea as a substitute for deletion. If this method can work with 2 and 1 entries, then may be it's good enough. The manual one didn't.

**Reply by shmerl** [reply]
One possible way to improve the visual method of selection and to allow deleting all entries can be this:

* Add a way to differentiate sending an empty list (nothing selected) to jumping to the list entry. I.e. pressing Enter on a a list with no selections now just jumps to the entry without updating the list to empty.
* Handle selection of 1 out of 2 as well (it's ignoring it now).

Then all deletion scenarios will be essentially covered with the visual method.

**Reply by shmerl** [reply]
I think I was able to make the above snippet with Ctr+x work to remove one entry out of two, but it doesn't work for removing the last entry

**Reply by ibhagwan** [reply]
Obviously you cannot create a new list from no entries, why would that be important to “clear” the list? If you’re done with the list’s usefulness move on to the next list, what’s the point of deleting the last entry?

**Reply by shmerl** [reply]
Well, let's say I finished working on the list and now I want to work on something else (not related to quickfix list). Lingering last entry still shows up in the list for me, so I'd rather remove it. With dd method it can remove all of them.

It's somewhat surprising that default neovim's / vim's features for handling entries in that list are so limited and require literally calling an API function to manipulate instead of having some simple commands. I guess the list was never intended for direct use and rather is a tool for plugins to utilize.

**Comment by shmerl** [comment]
If anyone needs, I found another tip that helps with these kind of workflows.

Customize your ripgrep call like this (i.e. add `--sort=path` to however you are using `rg`):

```vim
:lua require('fzf-lua').live_grep_native({ cmd = "rg ... --sort=path" }, resume = true)
```

This will ensure result of rg is deterministic and sorted (at the cost of running it single threaded). This is super useful when re-running searches especially with `resume = true`). Non deterministic output of rg makes things very confusing when re-running. Note that without `--sort=path` rg is parallelized and faster. So use only when needed.
